modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTask.java

```diff
@@ -295,26 +300,104 @@ private static ElasticsearchException wrapException(Exception ex) {
+
+                    if (oldDoc.getAllocationId() == null || allocationId > oldDoc.getAllocationId()) {
+                        ReindexTaskStateDoc newDoc = new ReindexTaskStateDoc(request, allocationId, response, exception, failureStatusCode,
+                            checkpoint);
+                        reindexIndexClient.updateReindexTaskDoc(persistentTaskId, newDoc, term, seqNo, new ActionListener<>() {
+                            @Override
+                            public void onResponse(ReindexTaskState newTaskState) {
+                                assignmentListener.onAssignment(newTaskState);
+                            }
+
+                            @Override
+                            public void onFailure(Exception ex) {
+                                if (ex instanceof VersionConflictEngineException) {
+                                    // There has been an indexing operation since the GET operation. Try
+                                    // again if there are assignment attempts left.
+                                    if (assignmentAttempts < 3) {
+                                        assign(assignmentListener);
+                                    } else {
+                                        logger.info("Failed to write allocation id to reindex task doc after maximum retry attempts", ex);
+                                        assignmentListener.onFailure(ReindexJobState.Status.ASSIGNMENT_FAILED, ex);
+                                    }
+                                } else {
+                                    logger.info("Failed to write allocation id to reindex task doc", ex);
+                                    assignmentListener.onFailure(ReindexJobState.Status.FAILED_TO_WRITE_TO_REINDEX_INDEX, ex);
+                                }
+                            }
+                        });
+                    } else {
+                        ElasticsearchException ex = new ElasticsearchException("A newer task has already been allocated");
+                        assignmentListener.onFailure(ReindexJobState.Status.ASSIGNMENT_FAILED, ex);
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception ex) {
+                    logger.info("Failed to fetch reindex task doc", ex);
+                    assignmentListener.onFailure(ReindexJobState.Status.FAILED_TO_READ_FROM_REINDEX_INDEX, ex);
+                }
+            });
+        }
+
+        private interface Listener {
+
+            void onAssignment(ReindexTaskState reindexTaskState);
+
+            void onFailure(ReindexJobState.Status status, Exception exception);
+        }
+    }
+
+    private class ProgressState implements Reindexer.CheckpointListener {
 
-        CheckpointHandler(ReindexTaskIndexState lastState) {
-            this.lastState = lastState;
+        private final Semaphore semaphore = new Semaphore(1);
```

Comment:
Can we do this without the blocking nature of a semaphore (which is always dangerous)?

Reply:
I added a TODO that I can immediately follow-up with. But it will require (I think) scheduling a retry on a thread pool. So I kind of want that to be dedicated PR.